modules/mini-games/speedrun.lua

```diff
@@ -0,0 +1,537 @@
+local Task         = require 'utils.task'
+local TeamSelector = require 'modules.gui.mini_game_team_selector'
+
+local targets    = {}
+local lookups    = {}
+local primitives = {}
+local progress   = {}
+local surfaces   = {}
+local forces     = {}
+
+local goals = require 'config.mini_games.speedrun'
+for index, goal in ipairs(goals) do
+    local lookup = {}
+    lookups[index] = lookup
+    targets[index] = goal.name
+
+    local ctn = goal.rockets or 0
+    lookup.rockets = goal.rockets and goal.rockets > 0
+    for _, key in ipairs{'research', 'items', 'entities'} do
+        local value = goal[key]
+        if value then
+            ctn = ctn + #value
+            lookup[key] = true
+            for _, name in ipairs(value) do lookup[key..'/'..name] = true end
+        end
+    end
+
+    goal.total = ctn
+end
+
+Global.register({
+    primitives = primitives,
+    progress   = progress,
+    surfaces   = surfaces,
+    forces     = forces
+}, function(tbl)
+    primitives = tbl.primitives
+    progress   = tbl.progress
+    surfaces   = tbl.surfaces
+    forces     = tbl.forces
+end)
+
+----- Local Variables ----
+
+local starting_items = {
+    {name = 'iron-gear-wheel', count = 8},
+    {name = 'iron-plate', count = 16}
+}
+
+----- Local Functions -----
+
+--- Internal, Used to clear tables of all values
+local function reset_table(table)
+    for k in pairs(table) do
+        table[k] = nil
+    end
+end
+
+--- Internal, Reset all global tables
+local function reset_globals()
+    reset_table(primitives)
+    reset_table(progress)
+    reset_table(surfaces)
+    reset_table(forces)
+end
+
+----- Map Gen -----
+
+local function ready_condition()
+    local ctn = 0
+    for _ in pairs(surfaces) do ctn = ctn + 1 end
+    return ctn == primitives.team_count
+end
+
+local surface_generator = Token.register(function(remaining)
+    local last = #remaining
+    local next_surface = remaining[last]
+    if not next_surface then return false end
+    remaining[last] = nil
+
+    local surface = game.create_surface(next_surface, { seed=remaining.seed })
+    surface.request_to_generate_chunks({0, 0}, 5)
+    surfaces[next_surface] = surface
+
+    return true
+end)
+
+----- Game Init and Start -----
+
+--- First function called by the mini game core to prepare for the start of a game
+local team_entry, timer_container
+local function init(args)
+    local target = tonumber(args[1])
+    if not target or target < 1 or target > #targets then Mini_games.error_in_game('Target index out of range') end
+    primitives.target = target
+    primitives.lookup = lookups[target]
+
+    local team_count = tonumber(args[2])
+    if not team_count or team_count < 1 then Mini_games.error_in_game('Team count is invalid') end
+    primitives.team_count = team_count
+
+    -- Create a surface for each team with the same seed and settings
+    local remaining, indicators = { seed = math.random(4294967295) }, goals[target]
```

Comment:
mabye allow seed to be set in the config optionaly too as then we do the same run a few times

Reply:
Yeah I'll add this and just have blank input mean random.

Reply:
Has been added see below